File: src/agentscope/app/_router/_schema/_voice_profile.py
# -*- coding: utf-8 -*-
"""Schema models for the voice profile router."""
from pydantic import BaseModel

from ...storage._model import VoiceProfileData, VoiceProfileRecord


class CreateVoiceProfileRequest(BaseModel):
    """Request body for creating a voice profile."""

    data: VoiceProfileData
    """The voice profile configuration."""


class CreateVoiceProfileResponse(BaseModel):
    """Response body for creating a voice profile."""

    profile_id: str
    """The created profile's unique identifier."""


class ListVoiceProfilesResponse(BaseModel):
    """Response body for listing voice profiles."""

    profiles: list[VoiceProfileRecord]
    """All voice profiles for the user."""

    total: int
    """Total count of profiles."""


class UpdateVoiceProfileRequest(BaseModel):
    """Request body for updating a voice profile."""

    data: VoiceProfileData
    """The updated voice profile configuration."""


class EngineInfo(BaseModel):
    """Metadata about a TTS engine."""

    name: str
    """The engine name (e.g. 'dashscope_tts')."""

    source: str
    """Whether the engine runs in cloud or locally."""

    gpu_requirement: str | None = None
    """GPU requirement for local engines."""

    voice_cloning: bool = False
    """Whether the engine supports voice cloning."""


class AvailableEnginesResponse(BaseModel):
    """Response body for available TTS engines."""

    engines: list[str]
    """Sorted list of engine names with credentials."""

    engine_details: list[EngineInfo] = []
    """Engine metadata including source and cloning."""


class CloneVoiceRequest(BaseModel):
    """Request body for voice cloning."""

    engine: str
    """TTS engine to clone for."""

    model: str | None = None
    """Target model (required for DashScope, optional for OpenAI)."""

    audio_base64: str | None = None
    """Base64-encoded audio (for Qwen-TTS/OpenAI upload)."""

    audio_filename: str = "voice.wav"
    """Original filename (used for MIME detection)."""

    audio_url: str | None = None
    """Publicly accessible audio URL (for CosyVoice)."""

    text: str | None = None
    """Audio transcript (improves Qwen-TTS quality)."""

    prefix: str = "agentscope"
    """Prefix for the generated voice name."""

    consent: str | None = None
    """OpenAI consent token (required for OpenAI cloning)."""


class CloneVoiceResponse(BaseModel):
    """Response body for voice cloning."""

    voice_id: str
    """The cloned voice identifier."""


class OpenAIConsentRequest(BaseModel):
    """Request body for uploading an OpenAI consent recording."""

    name: str = "agentscope_consent"
    """Label for the consent recording."""

    language: str = "en-US"
    """BCP 47 language tag for the consent phrase."""

    audio_base64: str
    """Base64-encoded consent audio recording."""

    audio_filename: str = "consent.wav"
    """Original filename (used for MIME detection)."""


class OpenAIConsentResponse(BaseModel):
    """Response body for OpenAI consent upload."""

    consent_id: str
    """The consent recording identifier (cons_xxx)."""

File: examples/web_ui/frontend/src/api/voiceProfile.ts
import { client } from './client';

export interface VoiceProfileData {
	name: string;
	engine?: string | null;
	model?: string | null;
	source?: 'api' | 'local' | null;
	voice?: string | null;
	metadata?: Record<string, unknown> | null;
}

export interface VoiceProfileRecord {
	id: string;
	user_id: string;
	created_at: string;
	updated_at: string;
	data: VoiceProfileData;
}

export interface ListVoiceProfilesResponse {
	profiles: VoiceProfileRecord[];
	total: number;
}

export interface CreateVoiceProfileResponse {
	profile_id: string;
}

export interface EngineInfo {
	name: string;
	source: 'api' | 'local';
	gpu_requirement?: string | null;
	voice_cloning: boolean;
}

export interface AvailableEnginesResponse {
	engines: string[];
	engine_details: EngineInfo[];
}

export interface CloneVoiceRequest {
	engine: string;
	model?: string | null;
	audio_base64?: string | null;
	audio_filename?: string;
	audio_url?: string | null;
	text?: string | null;
	prefix?: string;
	consent?: string | null;
}

export interface CloneVoiceResponse {
	voice_id: string;
}

export interface OpenAIConsentRequest {
	name?: string;
	language?: string;
	audio_base64: string;
	audio_filename?: string;
}

export interface OpenAIConsentResponse {
	consent_id: string;
}

export const voiceProfileApi = {
	list: () => client.get<ListVoiceProfilesResponse>('/voice-profile/'),

	get: (profileId: string) => client.get<VoiceProfileRecord>(`/voice-profile/${profileId}`),

	create: (data: VoiceProfileData) =>
		client.post<CreateVoiceProfileResponse>('/voice-profile/', { data }),

	update: (profileId: string, data: VoiceProfileData) =>
		client.patch<VoiceProfileRecord>(`/voice-profile/${profileId}`, { data }),

	delete: (profileId: string) => client.delete(`/voice-profile/${profileId}`),

	availableEngines: () =>
		client.get<AvailableEnginesResponse>('/voice-profile/available-engines'),

	cloneVoice: (data: CloneVoiceRequest) =>
		client.post<CloneVoiceResponse>('/voice-profile/clone', data),

	uploadOpenAIConsent: (data: OpenAIConsentRequest) =>
		client.post<OpenAIConsentResponse>('/voice-profile/openai-consent', data),
};
